# 데이터 전처리 및 QA

전처리 전 원본 데이터 오류 확인 후 처리
- 원본 : 1,465건
- 최종 : 1,212건(253건 중복 제거)

In [39]:
import pandas as pd
import numpy as np

df = pd.read_csv('amazon.csv')  
print(df.shape)
df.head()

(1465, 16)


,product_id,product_name,category,discounted_price,actual_price,discount_percentage,rating,rating_count,about_product,user_id,user_name,review_id,review_title,review_content,img_link,product_link
0,B07JW9H4J1,Wayona Nylon Braided USB to Lightning Fast Cha...,Computers&Accessories|Accessories&Peripherals|...,₹399,"₹1,099",64%,4.2,"24,269",High Compatibility : Compatible With iPhone 12...,"AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...","Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...","R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1K...","Satisfied,Charging is really fast,Value for mo...",Looks durable Charging is fine tooNo complains...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...
1,B098NS6PVG,Ambrane Unbreakable 60W / 3A Fast Charging 1.5...,Computers&Accessories|Accessories&Peripherals|...,₹199,₹349,43%,4.0,"43,994","Compatible with all Type C enabled devices, be...","AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...","ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Plac...","RGIQEG07R9HS2,R1SMWZQ86XIN8U,R2J3Y1WL29GWDE,RY...","A Good Braided Cable for Your Type C Device,Go...",I ordered this cable to connect my phone to An...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Ambrane-Unbreakable-Char...
2,B096MSW6CT,Sounce Fast Phone Charging Cable & Data Sync U...,Computers&Accessories|Accessories&Peripherals|...,₹199,"₹1,899",90%,3.9,"7,928",【 Fast Charger& Data Sync】-With built-in safet...,"AGU3BBQ2V2DDAMOAKGFAWDDQ6QHA,AESFLDV2PT363T2AQ...","Kunal,Himanshu,viswanath,sai niharka,saqib mal...","R3J3EQQ9TZI5ZJ,R3E7WBGK7ID0KV,RWU79XKQ6I1QF,R2...","Good speed for earlier versions,Good Product,W...","Not quite durable and sturdy,https://m.media-a...",https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Sounce-iPhone-Charging-C...
3,B08HDJ86NZ,boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...,Computers&Accessories|Accessories&Peripherals|...,₹329,₹699,53%,4.2,"94,363",The boAt Deuce USB 300 2 in 1 cable is compati...,"AEWAZDZZJLQUYVOVGBEUKSLXHQ5A,AG5HTSFRRE6NL3M5S...","Omkar dhale,JD,HEMALATHA,Ajwadh a.,amar singh ...","R3EEUZKKK9J36I,R3HJVYCLYOY554,REDECAZ7AMPQC,R1...","Good product,Good one,Nice,Really nice product...","Good product,long wire,Charges good,Nice,I bou...",https://m.media-amazon.com/images/I/41V5FtEWPk...,https://www.amazon.in/Deuce-300-Resistant-Tang...
4,B08CF3B7N1,Portronics Konnect L 1.2M Fast Charging 3A 8 P...,Computers&Accessories|Accessories&Peripherals|...,₹154,₹399,61%,4.2,"16,905",[CHARGE & SYNC FUNCTION]- This cable comes wit...,"AE3Q6KSUK5P75D5HFYHCRAOLODSA,AFUGIFH5ZAFXRDSZH...","rahuls6099,Swasat Borah,Ajay Wadke,Pranali,RVK...","R1BP4L2HH9TFUP,R16PVJEXKV6QZS,R2UPDB81N66T4P,R...","As good as original,Decent,Good one for second...","Bought this instead of original apple, does th...",https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Portronics-Konnect-POR-1...


In [40]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1465 entries, 0 to 1464
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   product_id           1465 non-null   object
 1   product_name         1465 non-null   object
 2   category             1465 non-null   object
 3   discounted_price     1465 non-null   object
 4   actual_price         1465 non-null   object
 5   discount_percentage  1465 non-null   object
 6   rating               1465 non-null   object
 7   rating_count         1463 non-null   object
 8   about_product        1465 non-null   object
 9   user_id              1465 non-null   object
 10  user_name            1465 non-null   object
 11  review_id            1465 non-null   object
 12  review_title         1465 non-null   object
 13  review_content       1465 non-null   object
 14  img_link             1465 non-null   object
 15  product_link         1465 non-null   object
dtypes: obj

## 데이터 오류 탐지

In [41]:
error_log = {}

In [42]:
rating_errors = pd.to_numeric(df['rating'], errors='coerce').isna().sum()

error_log['형식오류_rating'] = rating_errors

print(f"형식오류_rating: {rating_errors}건")

형식오류_rating: 1건


In [43]:
price_errors = df['discounted_price'].astype(str).str.contains('₹').sum()

error_log['형식오류_가격'] = price_errors

print(f"형식오류_가격: {price_errors}건")

형식오류_가격: 1465건


In [44]:
discount_errors = df['discount_percentage'].astype(str).str.contains('%').sum()

error_log['형식오류_할인율'] = discount_errors

print(f"형식오류_할인율: {price_errors}건")

형식오류_할인율: 1465건


In [45]:
duplicate_errors = df.duplicated(subset=['review_title', 'review_content']).sum()

error_log['중복오류_리뷰'] = duplicate_errors

print(f"중복 리뷰: {duplicate_errors}건")

중복 리뷰: 253건


In [46]:
print(error_log)

{'형식오류_rating': np.int64(1), '형식오류_가격': np.int64(1465), '형식오류_할인율': np.int64(1465), '중복오류_리뷰': np.int64(253)}


In [47]:
report = pd.DataFrame.from_dict(error_log, orient='index', columns=['건수'])
report

,건수
형식오류_rating,1
형식오류_가격,1465
형식오류_할인율,1465
중복오류_리뷰,253


In [48]:
report['비율(%)'] = (qa_report['건수'] / len(df) * 100).round(1)
report

,건수,비율(%)
형식오류_rating,1,0.1
형식오류_가격,1465,100.0
형식오류_할인율,1465,100.0
중복오류_리뷰,253,17.3


- 형식오류_가격 / 형식오류_할인율(100%)
  → 데이터 수집 단계에서 문자열로 저장됨
  → 전처리 필수, 수집 파이프라인 개선 필요
- 형식오류_rating: 1건 (0.1%)
  → "|" 값 입력 오류
- 중복오류_리뷰: 253건 (17.3%)
  → 리뷰 매크로 또는 크롤링 중복 수집으로 판단

## 전처리 - 가격

In [49]:
df['discounted_price'] = (
    df['discounted_price']
    .str.replace('₹', '')
    .str.replace(',', '')
    .str.strip()
    .astype(float)
)

df['actual_price'] = (
    df['actual_price']
    .str.replace('₹', '')
    .str.replace(',', '')
    .str.strip()
    .astype(float)
)


df[['discounted_price', 'actual_price']].describe()

,discounted_price,actual_price
count,1465.000000,1465.000000
mean,3125.310874,5444.990635
std,6944.304394,10874.826864
min,39.000000,39.000000
25%,325.000000,800.000000
50%,799.000000,1650.000000
75%,1999.000000,4295.000000
max,77990.000000,139900.000000


## 전처리 - 할인율

In [50]:
df['discount_percentage'] = (
    df['discount_percentage']
    .str.replace('%', '')
    .str.strip()
)
df['discount_percentage'] = pd.to_numeric(df['discount_percentage'], errors='coerce')

print(df['discount_percentage'].head())

0    64
1    43
2    90
3    53
4    61
Name: discount_percentage, dtype: int64


## 가격 구간 만들기

In [51]:
df['discount_amount'] = df['actual_price'] - df['discounted_price']

df[['actual_price', 'discounted_price', 'discount_amount']].head()

,actual_price,discounted_price,discount_amount
0,1099.0,399.0,700.0
1,349.0,199.0,150.0
2,1899.0,199.0,1700.0
3,699.0,329.0,370.0
4,399.0,154.0,245.0


In [52]:
bins   = [0, 500, 1000, 3000, 10000, 9999999]
labels = ['~500', '500~1K', '1K~3K', '3K~10K', '10K+']

df['price_tier'] = pd.cut(df['discounted_price'], bins=bins, labels=labels)

ratio1 = (df['price_tier'].value_counts(normalize = True) * 100).round(2)
ratio1

price_tier
~500      39.32
1K~3K     25.26
500~1K    17.20
3K~10K     9.90
10K+       8.33
Name: proportion, dtype: float64

전체 상품의 약 39%가 저가 상품에 집중되어 있음, 그러나 매출 기여도에서는 고가 구간 상품이 1건당 금액의 크기 때문에 더 높을 수 있음.

## 할인율 구간 만들기

In [53]:
disc_bins   = [0, 20, 40, 60, 80, 100]
disc_labels = ['0~20%', '21~40%', '41~60%', '61~80%', '81~100%']

df['discount_tier'] = pd.cut(df['discount_percentage'], bins=disc_bins, labels=disc_labels)

ratio2 = (df['discount_tier'].value_counts(normalize=True).sort_index() * 100).round(1)
ratio2

discount_tier
0~20%       9.3
21~40%     23.9
41~60%     35.9
61~80%     26.8
81~100%     4.2
Name: proportion, dtype: float64

- 전체 상품의 약 36%가 41%-60%에 집중되어 있음.
- 아마존은 중간 할인율 상품이 주력임을 알 수 있음.

## 평점 구간 만들기

In [54]:
df['rating'] = pd.to_numeric(df['rating'], errors='coerce')
df['rating'] = df['rating'].fillna(df['rating'].mean().round(1))

이상값 발견 '|' -> pd.to_numeric으로 처리

In [55]:
rating_bins   = [0, 3.0, 3.5, 4.0, 4.5, 5.0]
rating_labels = ['~3.0', '3.0~3.5', '3.5~4.0', '4.0~4.5', '4.5~5.0']

df['rating_tier'] = pd.cut(df['rating'], bins=rating_bins, labels=rating_labels)

ratio3 = (df['rating_tier'].value_counts(normalize=True).sort_index() * 100).round(1)
ratio3

rating_tier
~3.0        0.7
3.0~3.5     4.0
3.5~4.0    31.9
4.0~4.5    61.5
4.5~5.0     2.0
Name: proportion, dtype: float64

전체 상품 61.5%가 4.0-4.5에 몰려있음 -> 평점이 전반적으로 높지만 신뢰도 이슈가 있을 수 있음, 이후 평점별 구매 전환율 차이 확인 필요

## 카테고리 파싱

In [56]:
df['category_main'] = df['category'].str.split('|').str[0]
df['category_sub']  = df['category'].str.split('|').str[1]

df[['category', 'category_main', 'category_sub']].head(3)

,category,category_main,category_sub
0,Computers&Accessories|Accessories&Peripherals|...,Computers&Accessories,Accessories&Peripherals
1,Computers&Accessories|Accessories&Peripherals|...,Computers&Accessories,Accessories&Peripherals
2,Computers&Accessories|Accessories&Peripherals|...,Computers&Accessories,Accessories&Peripherals


## 중복 리뷰 제거

In [57]:
before = len(df)
df = df.drop_duplicates(subset=['review_title', 'review_content'])
after = len(df)

print(f"제거 전: {before}건 → 제거 후: {after}건 ({before - after}건 제거)")

제거 전: 1465건 → 제거 후: 1212건 (253건 제거)


In [58]:
df.to_csv('amazon_cleaned.csv', index=False)
print("저장 완료!")

저장 완료!


## 전처리 완료

원본 1,465건에서 중복 리뷰 253건을 제거하여
최종 1,212건의 데이터를 저장.

<처리한 오류>
- 가격/할인율 특수문자 제거 (₹, %, 콤마)
- rating 이상값 처리 ("|" → 평균값 대체)
- 중복 리뷰 253건 제거